### PR TITLE
Price Discovery SC

### PR DESCRIPTION
**Pull request by dorin-iancu**

### CODE REVIEW

File: price-discovery/src/create_pool.rs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        let (lp_token, _, _) = contract_call.execute_on_dest_context().into_tuple();
+        self.lp_token_id().set(&lp_token.token_identifier);
+        self.total_lp_tokens_received().set(&lp_token.amount);
+
+        Ok(())
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn accept_funds_func(&self) -> SCResult<()> {
```
Comment (sasurobert) [inline price-discovery/src/create_pool.rs:76]: maybe add a check that this endpoint is called only once.

File: price-discovery/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        self.accepted_token_id().set(&accepted_token_id);
+        self.start_epoch().set(&start_epoch);
+        self.end_epoch().set(&end_epoch);
+
+        Ok(())
+    }
+
+    #[only_owner]
+    #[payable("*")]
+    #[endpoint(depositInitialTokens)]
```
Comment (sasurobert) [inline price-discovery/src/lib.rs:63]: why is this needed ? like initial depoist ?

Comment (sasurobert) [reply price-discovery/src/lib.rs:63]: and why only owner ?

File: price-discovery/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+        let caller = self.blockchain().get_caller();
+        self.mint_and_send_redeem_token(&caller, LAUNCHED_TOKEN_REDEEM_NONCE, &payment_amount);
+
+        Ok(())
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn deposit(
```
Comment (sasurobert) [inline price-discovery/src/lib.rs:87]: I think only one deposit is enough. And according to tokenID you go for the tokenA or tokenB nonce (initial token or launchpad token).

File: price-discovery/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        &self,
+        #[payment_token] payment_token: TokenIdentifier,
+        #[payment_nonce] payment_nonce: u64,
+        #[payment_amount] payment_amount: BigUint,
+    ) -> SCResult<()> {
+        self.require_active()?;
+
+        let redeem_token_id = self.redeem_token_id().get();
+        require!(
+            payment_token == redeem_token_id && payment_nonce == ACCEPTED_TOKEN_REDEEM_NONCE,
```
Comment (sasurobert) [inline price-discovery/src/lib.rs:115]: withdraw can be done for accepted_token and launchpad_token as well. Some users have tokenA, others have tokenB they can simply enter with either of them. As the price discovery goes on, some users might want to withdraw their position - because they do not like how the price will be once the LP is launch. They can withdraw tokenA or tokenB.

File: price-discovery/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        let start_epoch = self.start_epoch().get();
+        let end_epoch = self.end_epoch().get();
+        require!(
+            start_epoch <= current_epoch,
+            "Deposit period not started yet"
+        );
+        require!(current_epoch < end_epoch, "Deposit period ended");
+
+        let launched_token_id = self.launched_token_id().get();
+        let current_launched_token_balance =
```
Comment (sasurobert) [inline price-discovery/src/lib.rs:169]: no need for this check as there is no need for the initial deposit.

File: price-discovery/src/redeem_token.rs [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+pub const LAUNCHED_TOKEN_REDEEM_NONCE: u64 = 1;
+pub const ACCEPTED_TOKEN_REDEEM_NONCE: u64 = 2;
+
+#[elrond_wasm::module]
+pub trait RedeemTokenModule {
+    #[only_owner]
+    #[payable("EGLD")]
+    #[endpoint(issueRedeemToken)]
+    fn issue_redeem_token(
```
Comment (sasurobert) [inline price-discovery/src/redeem_token.rs:11]: please integrate the new function which does everything in ONE go, issue esdt and local roles

File: price-discovery/src/create_pool.rs (hunk 1 above)
Comment (dorin-iancu) [reply price-discovery/src/create_pool.rs:76]: This is basically just a callback for the dex `add_initial_liquidity`, so it can't be called directly. Additionally, it does not alter the state, so even if someone managed to somehow trick the DEX contracts to call this endpoint, it wouldn't affect anything.

File: price-discovery/src/lib.rs (hunk 2 above)
Comment (dorin-iancu) [reply price-discovery/src/lib.rs]: Removed, now only the `deposit` endpoint exists.

File: price-discovery/src/lib.rs (hunk 3 above)
Comment (dorin-iancu) [reply price-discovery/src/lib.rs:63]: Done.

File: price-discovery/src/lib.rs (hunk 4 above)
Comment (dorin-iancu) [reply price-discovery/src/lib.rs]: Done.

File: price-discovery/src/lib.rs (hunk 5 above)
Comment (dorin-iancu) [reply price-discovery/src/lib.rs]: Removed.

File: price-discovery/src/redeem_token.rs (hunk 6 above)
Comment (dorin-iancu) [reply price-discovery/src/redeem_token.rs:35]: Done.

File: price-discovery/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    fn compute_lp_amount_to_send(
+        &self,
+        redeem_token_nonce: u64,
+        redeem_token_amount: BigUint,
+    ) -> BigUint {
+        let total_lp_tokens = self.total_lp_tokens_received().get();
+
+        match redeem_token_nonce {
```
Comment (sasurobert) [inline price-discovery/src/lib.rs:163]: we need to make a common code with swap here - as this is critical to be the same. Right now it looks ok.

File: price-discovery/src/redeem_token.rs [hunk 8]
```diff
@@ -0,0 +1,123 @@
+elrond_wasm::imports!();
+use hex_literal::hex;
+
+pub const LAUNCHED_TOKEN_REDEEM_NONCE: u64 = 1;
+pub const ACCEPTED_TOKEN_REDEEM_NONCE: u64 = 2;
+
+const SFT_TOKEN_TYPE_NAME: &[u8] = b"SFT";
```
Comment (sasurobert) [inline price-discovery/src/redeem_token.rs:7]: META type has to be used. this is a financial token

Comment (dorin-iancu) [reply price-discovery/src/redeem_token.rs]: Done.

File: price-discovery/src/lib.rs (hunk 7 above)
Comment (dorin-iancu) [reply price-discovery/src/lib.rs:166]: Done. Now it's using the `quote` function from the `amm` module: https://github.com/ElrondNetwork/sc-dex-rs/blob/main/dex/pair/src/amm.rs#L16

**Review by sasurobert — APPROVED** [review]

File: price-discovery/src/lib.rs (hunk 7 above)
Comment (dorin-iancu) [reply price-discovery/src/lib.rs:164]: Changed to use `rule_of_three` instead of `quote`

File: dex/pair/src/fee.rs [hunk 9]
```diff
@@ -109,9 +109,10 @@ pub trait FeeModule:
     }
 
     fn reinject(&self, token: &TokenIdentifier, amount: &BigUint) {
-        let mut reserve = self.pair_reserve(token).get();
-        reserve += amount;
-        self.pair_reserve(token).set(&reserve);
+        self.pair_reserve(token)
+            .update(|reserve| *reserve += amount);
+        self.pair_virtual_reserve(token, token)
```
Comment (sasurobert) [inline dex/pair/src/fee.rs:114]: This is not needed - you need to redo the PR as you brought stuff from the main branch - which is not yet needed

File: dex/pair/src/liquidity_pool.rs [hunk 10]
```diff
@@ -189,6 +225,16 @@ pub trait LiquidityPoolModule:
         self.calculate_k_constant(&first_token_amount, &second_token_amount)
     }
 
+    fn calculate_k_for_virtual_reserves(&self, token_side_id: &TokenIdentifier) -> BigUint {
```
Comment (sasurobert) [inline dex/pair/src/liquidity_pool.rs:228]: these are functions from front-running-protection branch which was merged into main. We do not need these yet.

**Review by sasurobert — CHANGES_REQUESTED** [review]
